### PR TITLE
use LoadBalancer type service for e2e service test to patch ingress status

### LINKED ISSUE #119452: Failure cluster [878e052f...] ipMode: Required value: must be specified when `ip` is set
##### Error text:
```
[FAILED] Could not patch service status%!(EXTRA *errors.StatusError=Service "test-service-sq72c" is invalid: status.loadBalancer.ingress[0].ipMode: Required value: must be specified when `ip` is set): Service "test-service-sq72c" is invalid: status.loadBalancer.ingress[0].ipMode: Required value: must be specified when `ip` is set
In [It] at: test/e2e/network/service.go:3406 @ 07/17/23 21:30:54.906

```
#### Recent failures:
[2023/7/20 05:20:04 ci-kubernetes-e2e-gci-gce-alpha-enabled-default](https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/ci-kubernetes-e2e-gci-gce-alpha-enabled-default/1681775912099516416)
[2023/7/20 04:28:04 ci-kubernetes-e2e-gci-gce-alpha-enabled-default](https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/ci-kubernetes-e2e-gci-gce-alpha-enabled-default/1681762825808121856)
[2023/7/20 03:43:05 ci-kubernetes-e2e-gci-gce-alpha-enabled-default](https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/ci-kubernetes-e2e-gci-gce-alpha-enabled-default/1681751501241847808)
[2023/7/20 03:00:04 ci-kubernetes-e2e-gci-gce-alpha-enabled-default](https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/ci-kubernetes-e2e-gci-gce-alpha-enabled-default/1681740679534874624)
[2023/7/20 02:12:04 ci-kubernetes-e2e-gci-gce-alpha-enabled-default](https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/ci-kubernetes-e2e-gci-gce-alpha-enabled-default/1681728600425893888)

/kind failing-test

/sig network
https://github.com/kubernetes/kubernetes/pull/118895 seems to be related.

### PR DESCRIPTION
**Pull request by pacoxu**
#### What type of PR is this?
/kind failing-test
/kind flake

#### Which issue(s) this PR fixes:

Fixes #119452

#### Special notes for your reviewer:
The failure is after #118895. 
/cc @RyanAoh

This test is flake not failing every time.

#### Does this PR introduce a user-facing change?

```release-note
None
```

### CODE REVIEW

**Comment by robscott** [comment]
@RyanAoh It seems like there must be some kind of mistake here. If I'm understanding this PR correctly, it appears that `IPMode` was added as a required field without a default value, which would be a backwards incompatible change. Did you mean to add the new field as optional?

/cc @aojea @thockin

**Comment by RyanAoh** [comment]
`IPMode` is an optional field. Looks like we should allow the load-balancer status updated with only ingress.ip specified, and the ingress.ipMode will default to "VIP" eventually.
The fllowing validation may need to be removed：
https://github.com/kubernetes/kubernetes/blob/4457f85eb3dfa34e46d6d776a3b7e89088dc9279/pkg/apis/core/validation/validation.go#L7066-L7069
Any advice? @thockin

**Comment by pacoxu** [comment]
> The following validation may need to be removed.

- remove the validation 
- default to "VIP"

Either can solve the failure without the e2e test change.

**Comment by pacoxu** [comment]
The failed CI can be fixed by https://github.com/kubernetes/kubernetes/pull/119486. (Not related to this PR)

**Review by thockin — COMMENTED** [review]
Thanks!

/lgtm
/approve

**Comment by pacoxu** [comment]
/retest

Should this be merged in v1.28? If yes, we need to tag the milestone v1.28.

**Comment by aojea** [comment]
> Thanks!
> 
> /lgtm /approve

/hold

I'm not sure this is correct, @thockin does this mean that old clients will not work once this feature is enabled if they don't set the field?

**Comment by pacoxu** [comment]
> I'm not sure this is correct, @thockin does this mean that old clients will not work once this feature is enabled if they don't set the field?

https://github.com/kubernetes/kubernetes/blob/97c7dbcd227ec253bb277a91575706fe30f519dc/pkg/apis/core/v1/defaults.go#L146-L156

There is a logic to default IPMode to VIP.

**Comment by thockin** [comment]
The problem with tests is that they don't run thru the WHOLE API stack, so default values are not set.  We get away with it a lot, but then you hit something like this and you need to go touch all of the test cases.  This is why pkg/api/service/testing/ exists, but we still have to convert everything to use it (and we have to deal with versioned vs unversioned)

**Review by thockin — COMMENTED** [review]
/lgtm

**Comment by aojea** [comment]
> The problem with tests is that they don't run thru the WHOLE API stack, so default values are not set.

This is an e2e test, this runs thru the whole API stack

**Comment by pacoxu** [comment]
The validation

https://github.com/kubernetes/kubernetes/blob/4457f85eb3dfa34e46d6d776a3b7e89088dc9279/pkg/apis/core/validation/validation.go#L7066-L7069

The defaulting and test

https://github.com/kubernetes/kubernetes/blob/97c7dbcd227ec253bb277a91575706fe30f519dc/pkg/apis/core/v1/defaults.go#L146-L156

https://github.com/kubernetes/kubernetes/blob/0fdb2a07f8c4433db98a5e6247640221663bcc42/pkg/apis/core/v1/defaults_test.go#L1246-L1259

@aojea It sounds like that we should add a defaulting logic for service status. Right? I opened https://github.com/kubernetes/kubernetes/pull/119674 to do a test.

**Comment by thockin** [comment]
Antonio is right. I skipped that this is E2E. It should be defaulted.

**Comment by pacoxu** [comment]
The defaulting logic is only for `LoadBalancer`.

But the e2e is using a clusterIP.

I think this is the reason ?

**Comment by thockin** [comment]
Ahh.  The fact that we allow `ip` to be set when it's not a load-balancer is unfortunate.  I think the defaulting should change, so it sets `ipMode` whenever `ip` is set.

Hold on.  I see this code in pkg/registry/core/service/strategy.go: `dropTypeDependentFields()`:

```
    // Clear the load-balancer status if it is no longer appropriate.  Although
    // LB de-provisioning is actually asynchronous, we don't need to expose the
    // user to that complexity.
    if newSvc.Spec.Type != api.ServiceTypeLoadBalancer {
        newSvc.Status.LoadBalancer = api.LoadBalancerStatus{}
    }
```

Yet I just manually added an IP to a clusterIP service:

```
$ k get svc iamyouare -oyaml
apiVersion: v1
kind: Service
metadata:
  creationTimestamp: "2023-01-17T17:55:27Z"
  labels:
    app: iamyouare
    foo: bar2
  name: iamyouare
  namespace: default
  resourceVersion: "35550324"
  uid: 1787eb65-4d45-4ad7-856c-73fbeb274a3f
spec:
  clusterIP: 10.0.168.22
  clusterIPs:
  - 10.0.168.22
  internalTrafficPolicy: Cluster
  ipFamilies:
  - IPv4
  ipFamilyPolicy: SingleStack
  ports:
  - name: p1
    port: 9376
    protocol: TCP
    targetPort: 9376
  sessionAffinity: None
  type: ClusterIP
status:
  loadBalancer:
    ingress:
    - ip: 1.2.3.4
```

Why is that not being wiped?

**Comment by thockin** [comment]
Ugh, spec vs. status.

I think this may be one of the few cases where we can actually change the behavior in a "breaking" way.

User creates service with `.type=ClusterIP`, `.status.loadBalancer` is empty
Controller or user updates the status subresource, sets `.spec.loadBalancer.ingress[].ip` (not wiped or validated, should be)
User updates `service spec` in any way, `.staus.loadBalancer` is wiped

It's clear that the `.status.loadBalancer` is not meant to be set when `.type` is not "LoadBalancer".

The question is:

1) Fail validation?
2) Silently discard the value?
3) Warn and then discard the value?

Because we allowed it in the past, I lean towards 3.

@aojea and @khenidak - the people most familair with this type.

**Comment by thockin** [comment]
/lgtm

https://github.com/kubernetes/kubernetes/issues/119700 for followup

**Comment by pacoxu** [comment]
+1 for `Warn and then discard the value`. It seems to be a more API-consistent change.

**Comment by aojea** [comment]
/approve

heh, this is why everybody loves Service 🙃 

what an interesting failure

**Comment by aojea** [comment]
/hold cancel
/milestone next-candidate

only impacts an alpha feature gate

**Review by thockin — COMMENTED** [review]
I think the timeout at L3518 is now too short for an LB service.

We should either call GetServiceLoadBalancerPropagationTimeout or call e2eservice.WaitForServiceDeletedWithFinalizer or just bump this timeout or something.  

Do we really need to test patch here?  If not, it would be simpler to remove this section - it is exercised in the loadbalancer tests already.

@aojea

**Comment by aojea** [comment]
/hold

we need to understand the consequnces of using a LoadBalancer

File: test/e2e/network/service.go [hunk 1]
```diff
@@ -3350,7 +3350,7 @@ var _ = common.SIGDescribe("Services", func() {
 				Labels: testSvcLabels,
 			},
 			Spec: v1.ServiceSpec{
-				Type: "ClusterIP",
+				Type: "LoadBalancer",
```
Comment (aojea) [inline test/e2e/network/service.go:3353]: There is a trick @pacoxu , set the `spec.service.Spec.LoadBalancerClass` to something as `fakeClass` so the controllers on the cloud-provider skip it

Comment (thockin) [reply test/e2e/network/service.go:3353]: ooh, good idea

File: test/e2e/network/service.go [hunk 2]
```diff
@@ -3357,6 +3357,7 @@ var _ = common.SIGDescribe("Services", func() {
 					Port:       int32(80),
 					TargetPort: intstr.FromInt(80),
 				}},
+				LoadBalancerClass: utilpointer.String("test.com/test"),
```
Comment (thockin) [inline test/e2e/network/service.go:3360]: super super nit:  `example.com` should be used for things like this.

**Review by thockin — COMMENTED** [review]
I'm going to approve it, but not LGTM.  Change the domain name and then LGTM is easy.

File: test/e2e/network/service.go (hunk 2 above)
Comment (pacoxu) [reply test/e2e/network/service.go]: Updated to `"example.com/internal-vip`. Is this OK?

**Pull request merged**